### PR TITLE
audio messages

### LINKED ISSUE #10669: Audio Messages 🗣️💬
### User Story

As a user I want to be able to compose and send audio messages in chat instead of having to type it with keyboard. I want message recipients to be able to play back my audio messages.

### Description

The scope of this issue is to add a new message type of an `Audio Message` to our chat and the interface to record one.

We should cap the maximum length of an audio message to something reasonable and test what's the best compression / payload to send the message with. These don't have to be high quality, a 128kb/s bitrate is already better than what you get on a phone. I assume we're reusing the same Waku logic as we did for images.

 The UI:

- we add a new button, placed on the last position for starting a new audio message
- when tapped we show a brief hint that to start recording you need to press and hold it

tooltip

- when press and held, we animate in the recording UI, it's composed of two parts: one that has the audio waveform with controls for playback and cancelling and the duration in seconds. The other for pausing / continuing recording (we record only when the button is pressed) and sending the message.

recording
- ideally you should be able to scrub the waveform to position the play head when not recording, that can be followed up as a nice to have in a separate issue
- tapping outside the recording toolbar, dismisses the recording
- sending the message renders the audio message on the chat thread as expected. inside the message, we place a waveform, a control to play/pause and a time indicator.

sending

Prototype 👉 https://framer.cloud/mnwpw
_Figma design 👉 [here](https://www.figma.com/file/aS1ct66VQ6V0cio7vSqS8UoG/Chat?node-id=9856%3A1578)

### Acceptance Criteria

You can send and play back audio messages in chats

### Notes

Thanks! 🙌

### PR DESCRIPTION
**Pull request by tbenr**
fixes #10669 

Audio message implementation started

Uses a fork of `react-native-audio-toolkit` https://github.com/tbenr/react-native-audio-toolkit/tree/v2.0.3-status-v6 which implements

* metering while recording
* audio playback of base64 inline audio URL (`data:audio/mp3;base64,UklGRhwMAABXQVZF...`)
* fix to various bugs

recording params are: aac, 22050khz, mono, 32kbs

### sender side:

* [x] audio recording
* [x] recording playback
* [x] send base64 audio message
* [x] recording pause
* [x] fix issue on android recording pause (need to cherrypick as audiotoolkit pr)
* [x] max 120sec recording
* [x] delete file on recording canceled\sent
* [x] code cleanup\refactoring
* [x] UI
* [x] change show\hide behaviour of cancel\send button
* [x] add remain awake module during recording

### receiver side (chat view)

* [x] receive base64 audio and prepare player
* [x] playback
* [x] issue (ios only?): audio length is unknown before first playback
* [x] seeking
* [x] code cleanup\refactoring
* [x] UI
* [x] slider: revert to default thumb and use tint color
* [x] slider: increase update freq or animate

status: ready

### CODE REVIEW

**Comment by tbenr** [comment]
i added a testing ui to play with audio formats for interoperability testing and audio quality\size balancing

File: mobile/js_files/package.json [hunk 1]
```diff
@@ -10,6 +10,7 @@
     "app:android": "react-native run-android"
   },
   "dependencies": {
+    "@react-native-community/audio-toolkit": "git+https://github.com/tbenr/react-native-audio-toolkit.git#v2.0.3-status",
```
Comment (jakubgs) [inline mobile/js_files/package.json]: Why do we need yet another fork of a package?

File: ios/Podfile.lock [hunk 2]
```diff
@@ -419,6 +421,7 @@ DEPENDENCIES:
   - React-RCTVibration (from `../node_modules/react-native/Libraries/Vibration`)
   - ReactCommon/callinvoker (from `../node_modules/react-native/ReactCommon`)
   - ReactCommon/turbomodule/core (from `../node_modules/react-native/ReactCommon`)
+  - "ReactNativeAudioToolkit (from `../node_modules/@react-native-community/audio-toolkit`)"
```
Comment (jakubgs) [inline ios/Podfile.lock:435]: Why is this the only one quoted?

**Comment by jakubgs** [comment]
Looks like there's an issue with your dependency:
```
error: cannot update ref 'refs/heads/v2.0.3-status': trying to write non-commit object 53e823d1bd0dffd7bf799ccfb8fccd3e4023d00f to branch 'refs/heads/v2.0.3-status'
```
Have you at any point updated the `v2.0.3-status` tag?

**Comment by tbenr** [comment]
> Looks like there's an issue with your dependency:
> 
> ```
> error: cannot update ref 'refs/heads/v2.0.3-status': trying to write non-commit object 53e823d1bd0dffd7bf799ccfb8fccd3e4023d00f to branch 'refs/heads/v2.0.3-status'
> ```
> 
> Have you at any point updated the `v2.0.3-status` tag?

I dont think so, we discussed earlier this point so i should had care of it.
when happend nix was also failing locally too

**Comment by jakubgs** [comment]
I will need to take a look and debug `yarn2nix` I guess.

**Comment by jakubgs** [comment]
I think I found the issue: https://github.com/NixOS/nix/issues/2385

According to a commit that closed that issue:
```
only non-annotated tags can be fetched
```
https://github.com/NixOS/nix/commit/ae244af242ca3621e5a3b9196f27d9fcbf297266

But that line isn't there in current version of docs:
https://nixos.org/nix/manual/#ssec-builtins

Which doesn't make much sense to me. I'll try to reproduce.

**Comment by jakubgs** [comment]
Yup, tested it with a simple Nix derivation:
```nix
(import <nixpkgs> {}).stdenv.mkDerivation {
  name = "test";
  src = builtins.fetchGit {
    url = "https://github.com/tbenr/react-native-audio-toolkit.git";
    ref = "v2.0.3-status";
  };
}
```
And it fails the same way:
```
 > nix-build --show-trace
error: cannot update ref 'refs/heads/v2.0.3-status': trying to write non-commit object 53e823d1bd0dffd7bf799ccfb8fccd3e4023d00f to branch 'refs/heads/v2.0.3-status'
error: while evaluating the attribute 'src' of the derivation 'test' at /home/sochan/soft/nix-fetchgit-test/default.nix:2:3:
program 'git' failed with exit code 1
```
But it fetches it without issues if I change the `ref` to be `v2.0.3`:
```
 > nix-build --show-trace
these derivations will be built:
  /nix/store/gm72l91mw8bd5wy9s84bzca4xi557xyd-test.drv
building '/nix/store/gm72l91mw8bd5wy9s84bzca4xi557xyd-test.drv'...
unpacking sources
unpacking source archive /nix/store/vg34709b4dgkz4hndq3gzhhk2f66ngyr-source
source root is source
...
```
And when I checkout the repo the issue is clearly that `v2.0.3-status` is an annotated tag:
```
 > git for-each-ref | grep v2.0.3
96197e3d6643d47b86267bcc6cb29cb26536e226 commit	refs/tags/v2.0.3
53e823d1bd0dffd7bf799ccfb8fccd3e4023d00f tag	refs/tags/v2.0.3-status
```

**Comment by jakubgs** [comment]
I created an issue about it: https://github.com/status-im/status-react/issues/10825
It looks like you must have used an `-a`/`--annotated` flag when creating the tag.

**Comment by tbenr** [comment]
> I created an issue about it: #10825
> It looks like you must have used an `-a`/`--annotated` flag when creating the tag.

yes i did. I'll change it if continues failing. Thanks!!!

**Comment by jakubgs** [comment]
Run `make lint`.

**Comment by tbenr** [comment]
> Run `make lint`.

yeah i'm trying to remember it on every commit

**Comment by tbenr** [comment]
> > Run `make lint`.
> 
> yeah i'm trying to remember it on every commit

ok, now i know that`make lint-fix` doesn't give you the same feedback of `make lint`.

File: ios/Podfile.lock (hunk 2 above)
Comment (tbenr) [reply ios/Podfile.lock:435]: i haven't edit this maually.. i just did `make shell` -> `yarn add ...` and `make nix-update-pods`
i think all @react-native-community modules has "@" so you have all them quoted

Comment (tbenr) [reply ios/Podfile.lock:435]: I pinged react-native-community lol.

Comment (jakubgs) [reply ios/Podfile.lock:435]: I see.

**Comment by tbenr** [comment]
@flexsurfer I need some help on how to create a view which goes over the tabbar. 

image

see the #10669 for more details.

any hint on where is the best place to insert this view?

**Comment by flexsurfer** [comment]
i know @Ferossgp is working on new chat input

**Comment by oskarth** [comment]
Small request - could you please rebase to latest master so this commit https://github.com/status-im/status-react/commit/b5fda1234e5447cbc626059dfc78fcfbabc9e926 gets included? This means we aren't running E2E tests on eth.prod cluster anymore, and since this happens on every push, the impact is quite big. This ensures we have more accurate metrics going forward from Jul 1 onward, which would be awesome. See https://discuss.status.im/t/user-growth-and-retention/1782 for more

**Comment by Ferossgp** [comment]
Hey @tbenr I'm doing the new chat input [figma](https://www.figma.com/file/aS1ct66VQ6V0cio7vSqS8UoG/Chat?node-id=7782%3A36147), and I can do it as a part of new component. If it works for you then, in this PR you can make the UI in input simpler, like having them inside the bar. Otherwise to have it overflow you will need to wrap this view https://github.com/tbenr/status-react/blob/audio-messages/src/status_im/ui/screens/chat/input/input.cljs#L87 in an other view, and as second children add the overflow component. Starting from v1.5 we will use new gesture-handler and animated from here https://github.com/status-im/status-react/blob/develop/src/quo/animated.cljs https://github.com/status-im/status-react/blob/develop/src/quo/gesture_handler.cljs

**Comment by errorists** [comment]
Screenshot 2020-07-01 at 09 45 05


@tbenr regarding tab bar, feel free to just use it as it's here, no need to place it above or anything. You can ping me in Figma with any questions, I updated it there too

**Comment by tbenr** [comment]
@oskarth going to rebase

@Ferossgp sounds great, thanks for the hints. i'll work using the new design from @errorists

**Comment by tbenr** [comment]
@Ferossgp another question:
do you have any guidelines on how to do animations via react-native-svg we are using. I found https://medium.com/@ethantran/animating-svg-in-react-native-cf1907831608 
Is it the way to go for you?

**Comment by flexsurfer** [comment]
@tbenr awesome job so far, are you in status app, would want to send you some SNT for a beer :)

**Comment by tbenr** [comment]
@flexsurfer you find me there 😄  ens: tbenr

**Comment by Ferossgp** [comment]
@tbenr In terms of the tutorial, the one you've linked uses the standard react-native API. The one we switched is [react-native-reanimated](https://github.com/software-mansion/react-native-reanimated). This is the most advanced example I've seen, which includes most of the SVG interpolations https://www.youtube.com/watch?v=gLopy2MCAqM and code https://github.com/wcandillon/can-it-be-done-in-react-native/tree/master/season3/src/LiquidSwipe 
https://wcandillon.github.io/react-native-redash/svg there are some utils functions.

If the animation is not interactive (do not depend on direct interactions) it is fine to use the react-native API.

**Comment by tbenr** [comment]
@errorists @Ferossgp I did a first animation while recording (look at the updated screenshot, https://github.com/status-im/status-react/pull/10810#issue-434186277) tested only on iOS sim. Feedbacks\comments? (you can try the PR once compiled)

**Comment by hesterbruikman** [comment]
@tbenr quick note to let you know that that bounty is underway! So sorry, hit some roadblocks in Gitcoin account management. Amazing work on this!

**Comment by Ferossgp** [comment]
Hey @tbenr! I've tested it and the animation has a fast speed which makes it hard to follow while :(. I've looked at how it is implemented in other apps, and most do not show waves on the recording, only a text or a simpler indicator. While the waves that you've implemented are used only as a "map" for already recorded audio. If @errorists agree, I would go also with a simpler UI, so we can have it faster shipped, and then add the improvements upon usage feedback.

**Comment by errorists** [comment]
yea, I'd tend to agree. maybe the musician in me got carried away or I just love the Apple Messages interface which is where I started using audio messages. Are you guys comfortable with doing gestures then, because the alternative would use those. Press and hold on record, swipe your finger left to cancel - cue WhatsApp / Signal

edit: regarding the waveform animation, it is a good start as visually we'd need to make the bars much thinner, like 3px each with no margins and this would force us to animate at a slower pace, where to fill the entire length of the waveform preview you'd need to record around 4 seconds of audio.

**Comment by tbenr** [comment]
@Ferossgp @errorists thanks for the feedback. Since the animation is fully parametric I adjusted it to have more bars so you can have a look. @errorists is it running on physical iOS then? on android it seems I need to tune slience threshold differently.

**Comment by errorists** [comment]
@tbenr no, I couldn't get it to record on an iPhone, I'll try again tomorrow with the build having more bars.

**Comment by errorists** [comment]
@tbenr could we go with something much simpler instead of the waveform, it's clear by now that it would require too much time from everyone involved to get right, like in the latest build it looked like the bars weren't really mapped 1:1 to what I was saying anymore.

Screenshot 2020-07-07 at 12 11 12

So instead, a regular menu similar to images or stickers, only even simpler. Sorry for having you chase this earlier, we gave it a nice try :)

**Comment by tbenr** [comment]
@errorists yeah, simpler at first round makes sense! Do we also want to remove playback before sending?

**Comment by errorists** [comment]
@tbenr yes, let's simplify the UX, other messengers likewise seem to not have that capability so we should be fine :) thanks, all of the Figmas are updated, lmk if you need help with the animations on buttons.

**Comment by tbenr** [comment]
@errorists one clarification on user interaction: user needs to leave the button pressed and then swipe right or left to send or cancel?
or they are just normal buttons?

**Comment by errorists** [comment]
@tbenr just press the button, it changes state from the one on the left to the one on the right and begins recording. Pressing it again pauses the recording, but you should be able to hit it again to continue. `X` cancels the recording and returns to default empty state, and upward arrow sends it and returns to default empty state.

**Comment by tbenr** [comment]
@errorists take a look to the video https://youtu.be/B3Lag5-Wr7s to have a check

**Comment by errorists** [comment]
@tbenr dude, this pulsating voice record button is awesome. this is seriously an ace design input I didn't think of myself 👏 What's left, I think you need to use the Play/Pause icons, align the timestamps so they are at the same Y position, use the specified colours for the sliders, or better yet unify the sliders on both platforms so that hey look like the one in Figma.

**Comment by errorists** [comment]
also I couldn't tell from the video, can you pause the recording by tapping the recording button and then continue where you left off by tapping it again?

**Comment by tbenr** [comment]
> also I couldn't tell from the video, can you pause the recording by tapping the recording button and then continue where you left off by tapping it again?

@errorists 
it is the missing thing to be implemented. One question, what about fading in-out in loop the timer while paused? Do we want to change something on the main button while paused (color?)?

**Comment by errorists** [comment]
@tbenr no, just use the same state as when it's off. so the button has two states, start record - recording. Pressing it while recording goes back to the 'start recording' state

**Comment by errorists** [comment]
also with the message, I think it would be preferable if you had the message span the maximum width a message can have, I believe we have margins on both sides that control that right @Ferossgp ?

**Comment by tbenr** [comment]
@errorists do we have the play\pause\loading-spinner icons for the message button?

**Comment by errorists** [comment]
we do, you'll find all icons [here](https://www.figma.com/file/cb4p8AxLtTF3q1L6JYDnKN15/Index?node-id=5026%3A3382)

Select the one you're looking for, make sure you have the 24x24 parent layer selected and find the export panel in the right sidebar

Screenshot 2020-07-09 at 18 43 22

Alternatively, feel free to use the ones I exported for you in [this dropbox folder](https://www.dropbox.com/sh/ootinn18xikmstg/AABHZHcI3-x_FNYrPvztENbSa?dl=0)

The spinner is a small default Activity Indicator.

**Comment by hesterbruikman** [comment]
That pulsing record button is awesome! Great work on this @tbenr! Seems to work well for me on Android, but not on iOS. App crashes when I tap the record button. Any thoughts on what might be happening there?

Samsung S10e with Android 10 
Fairphone 2 with Android 7
iPhone 8 with iOS 13.5.1

**Comment by tbenr** [comment]
@hesterbruikman thanks for the feedback :) @errorists already told me there is that issue on ios. Unfortunately i cant test on a real ios device, only on emulator. errorist sent me status log but it isn't meaningful, i need OS logs related to the crash.
btw i'll try to run on iphone 8 - 13.5.1 simulator to try to recreate.

there is also an issue when pause recording on Android, but there I can fix.

**Comment by tbenr** [comment]
@errorists @errorists i did some research and this could be related to permissions in the plist. Later today i'll push an update to check.

**Comment by tbenr** [comment]
@errorists @hesterbruikman can you try again?

**Comment by hesterbruikman** [comment]
Rockstar! That works 🙌

**Comment by jinhojang6** [comment]
Legend 👍

**Comment by tbenr** [comment]
@hesterbruikman @errorists @cammellos this latest push has UI and functionalities completed on both android and iOS.
I added a maximum recording limit to 120sec.

**Comment by errorists** [comment]
@tbenr the recording interface is a joy to use, great work! 🙏  
some ui bugs I found: on the slider playheads, the one you asked assets for, the shadow is clipped, meaning it's placed in a smaller container than what's needed. 
Screenshot 2020-07-15 at 09 14 15
My only other gripe is that the slider's position could be updated more smoothly / frequently, like adding some frames in between the steps to ease it out so it doesn't jump 8 pixels or so at once. Definitely not a blocker, other than that in my limited testing I didn't find anything wrong.

**Comment by errorists** [comment]
@tbenr having used it some more, I'm wondering wouldn't it be better to _not_ hide the cancel / send buttons in the paused state? so if you pause, we change the state only for the recording button, and only hide the cancel / send buttons when there's nothing to send or cancel anyway

**Comment by tbenr** [comment]
> @tbenr the recording interface is a joy to use, great work! 🙏
> some ui bugs I found: on the slider playheads, the one you asked assets for, the shadow is clipped, meaning it's placed in a smaller container than what's needed.
> Screenshot 2020-07-15 at 09 14 15

@errorists I cropped the image you sent me cose the padding was causing the head to stay away from the far ends. I probably cropped too aggressively cutting out the shadow. Maybe you can provide me some better ones. consider that the issue in particular is the horizontal padding.

> My only other gripe is that the slider's position could be updated more smoothly / frequently, like adding some frames in between the steps to ease it out so it doesn't jump 8 pixels or so at once. Definitely not a blocker, other than that in my limited testing I didn't find anything wrong.

Yes you're right. I can increase the frequency or maybe adding an animation transition to smooth it up.. let me play with it.

**Comment by tbenr** [comment]
> @tbenr having used it some more, I'm wondering wouldn't it be better to _not_ hide the cancel / send buttons in the paused state? so if you pause, we change the state only for the recording button, and only hide the cancel / send buttons when there's nothing to send or cancel anyway

Oky!

**Comment by errorists** [comment]
@tbenr can't we set overflow to visible on the slider head? this way the larger image could be visible in a container that's smaller. I have no idea how to do it differently, other than removing the shadow altogether which isn't great, would reduce affordance.

**Comment by tbenr** [comment]
@errorists another thing we need to address is the fact that, since the recording continue without leaving the finger on display, the display can poweroff and app can go in background. Maybe we need to implement a keep-awake while recording.

**Comment by errorists** [comment]
@tbenr good idea, we should as long as recording is on

**Comment by tbenr** [comment]
> @tbenr can't we set overflow to visible on the slider head? this way the larger image could be visible in a container that's smaller. I have no idea how to do it differently, other than removing the shadow altogether which isn't great, would reduce affordance.

react-native-slider is quite limited on that. you just provide an image. Try to make it as smallest as possible and let's try it out.

**Comment by errorists** [comment]
@tbenr how about setting a custom thumb tint color instead of using an image? looking at the example that's possible, 
Screenshot 2020-07-15 at 10 11 56

I honestly can't make it work good with a single image to use on both light/dark mode without any shadow, and if the shadow is made small, you would see the slider track peeking from underneath.

**Comment by tbenr** [comment]
> @tbenr how about setting a custom thumb tint color instead of using an image? looking at the example that's possible,
> Screenshot 2020-07-15 at 10 11 56
> 
> I honestly can't make it work good with a single image to use on both light/dark mode without any shadow, and if the shadow is made small, you would see the slider track peeking from underneath.

https://github.com/react-native-community/react-native-slider#thumbtintcolor

yeah that's a solution but consider that tint color there is supported only by android so we need to pass the images rendered with hardcoded colors

**Comment by errorists** [comment]
@tbenr u sure? it looks like it works on iOS too
Screenshot 2020-07-15 at 10 24 09

**Comment by tbenr** [comment]
> @tbenr u sure? it looks like it works on iOS too
> Screenshot 2020-07-15 at 10 24 09

uh!? ..maybe documentation is outdated.. let me check

**Comment by tbenr** [comment]
@errorists yes the doc is not updated, but tint works only default head, custom image + tint doesn't work.

**Comment by errorists** [comment]
@tbenr yea that's cool, I meant with the default head so it will be a bit larger on iOS and smaller on Android, it's the best we can do.

**Comment by hesterbruikman** [comment]
@tbenr @errorists What are things that remain for audio messages? Can we timebox improvements and start QA testing? Would like to see this merged as soon as possible

**Comment by tbenr** [comment]
@hesterbruikman i updated the todo list. I can close remaining points this evening. If UI will look ok i think QA can start tomorrow

**Comment by hesterbruikman** [comment]
Awesome, thanks @tbenr!

**Comment by tbenr** [comment]
@errorists this is how it looks like in ios with default head and tint

image

**Comment by errorists** [comment]
@tbenr yea a custom smaller view with rounded corners would be better but I'll live with it :) maybe they'll update it some day. on iOS I would increase the margin between the slider component and the timestamps by 4 since it's a bit tight now but it's already good, non critical and I don't want to hold you up anymore.

**Comment by tbenr** [comment]
@errorists @hesterbruikman finished!

**Comment by errorists** [comment]
scared

**Comment by errorists** [comment]
@tbenr feedback from @simonricoo who did some testing

> 1. (I haven't gave the permission to record audio yet) When I press play, the countdowns starts immediately -> Then Im asked to allow permission -> the countdown is still running in the BG

> 2. when I hold the "circle" to rewind in the recording few sec for instance, the icon stays "play" it should stay "pause" when I put my finger up from the screen

👀 https://cdn.discordapp.com/attachments/624634427930312718/733227871329583164/video0.mov

> also there is a weird Play/Pause icon glitch when I'm rewinding...

File: src/status_im/chat/models/input.cljs [hunk 3]
```diff
@@ -126,6 +126,15 @@
...
                                             :text         "Update to latest version to see a nice image here!"})))))
 
+(fx/defn send-audio-message
+  [cofx audio-path duration current-chat-id]
+  (when-not (string/blank? audio-path)
+    (chat.message/send-message cofx {:chat-id           current-chat-id
+                                     :content-type      constants/content-type-audio
+                                     :audio-path        audio-path
+                                     :audio-duration-ms duration
+                                     :text              "Update to latest version to listen to an audio here!"})))
```
Comment (cammellos) [inline src/status_im/chat/models/input.cljs]: `to an audio message` maybe?

**Review by cammellos — APPROVED** [review]
Truly amazing work @tbenr !
It all looks good, I have only a question,
what happens if the duration of the message does not correspond to the audio? (we can cap to a reasonable number in status-go, and you are granted that it's >= 0, but we don't cap the maximum)

**Comment by tbenr** [comment]
> Truly amazing work @tbenr !
> It all looks good, I have only a question,
> what happens if the duration of the message does not correspond to the audio? (we can cap to a reasonable number in status-go, and you are granted that it's >= 0, but we don't cap the maximum)

@cammellos thanks man!
A part of some incoherent UI, the only thing that can happen is a seek on media player to a timestamp grater than the end of the file. Which i think can only produce a seek error or a transparent seek to the end of the file. I'll try it out.
But in any case I think it's possible to add a check to avoid the out of bound seek...

File: src/status_im/chat/models/input.cljs (hunk 3 above)
Comment (tbenr) [reply src/status_im/chat/models/input.cljs]: and moreover those messages should be all in i18n :-)

File: src/status_im/ui/screens/chat/message/audio.cljs [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+(defn play-pause [{:keys [message-id state-ref] :as params} audio]
+  (if (not= message-id @current-player-message-id)
+    ;; player has audio from another message, we need to reload
+    (reload-player params
+                   audio
+                   ;; on-success: audio is loaded, do we have an existing value to seek to?
+                   #(if-some [seek-time (:seek-to-ms @state-ref)]
+                      ;; check seek time against real audio duration and play
+                      (let [checked-seek-time (min (audio/get-player-duration @player-ref) seek-time)]
```
Comment (tbenr) [inline src/status_im/ui/screens/chat/message/audio.cljs:145]: @cammellos added this check

**Comment by tbenr** [comment]
> @tbenr feedback from @simonricoo who did some testing
> 
> > 1. (I haven't gave the permission to record audio yet) When I press play, the countdowns starts immediately -> Then Im asked to allow permission -> the countdown is still running in the BG
> 
> > 2. when I hold the "circle" to rewind in the recording few sec for instance, the icon stays "play" it should stay "pause" when I put my finger up from the screen
> 
> 👀 https://cdn.discordapp.com/attachments/624634427930312718/733227871329583164/video0.mov
> 
> > also there is a weird Play/Pause icon glitch when I'm rewinding...

@errorists @simonricoo I improved 2. check the latest PR.
But its not clear to me 1. can you make a video for this one too?

**Comment by Ferossgp** [comment]
Hey @tbenr! I've merged now the #10873 and it changes a bit the input logic. After rebase you will need to enable https://github.com/status-im/status-react/blob/develop/src/status_im/ui/screens/chat/components/input.cljs#L154 this icon here, and delete the button and animation code for the audio panel. Now it is handled inside the input. Also, the height is set by the component which wraps the panels, so now you will need just to fill it with the content instead of fixed height.

**Comment by churik** [comment]
@tbenr amazing work! 👍 
### 1) Rebase and align with new chat input is needed
### 2) Preview of  audio messages on home screen "Update to latest version"
Expected to see smth like: 'Audio message'

IMAGE 2020-07-20 12:12:00
Same for push notifications.


### 3)  Endless popup "Maximum recording time is reached"
In case of recording > 2 mins, new popup "Maximum recording time is reached" appears every second.
Only one way to close it - restarting the app.

**Steps:**
- start 1-1 chat
- start recording
- record > 3 mins
**Expected behavior:** one popup "Maximum recording time is reached", can close it and send recording
**Actual behavior:** new "Maximum recording time is reached" is appearing every ms 
photo_2020-07-20 12 21 57

 ### 4)  Warn user if he switched to other views that recording will be not saved

In case of tapping navigation buttons inside app, leaving chat view or switching to keyboard/stickers/images recording will be lost.
IMO to have system popup to warn user that recording will be lost and suggest options to proceed or to cancel switching is required.

**Comment by tbenr** [comment]
> Hey @tbenr! I've merged now the #10873 and it changes a bit the input logic. After rebase you will need to enable https://github.com/status-im/status-react/blob/develop/src/status_im/ui/screens/chat/components/input.cljs#L154 this icon here, and delete the button and animation code for the audio panel. Now it is handled inside the input. Also, the height is set by the component which wraps the panels, so now you will need just to fill it with the content instead of fixed height.

Hi @Ferossgp I'm integrating new input but I see that sticker and audio icons are not shown simultaneously. To make speech icon to appear i need to comment out https://github.com/status-im/status-react/blob/e8fca7ce6877272ae51832939d032f04789bde22/src/status_im/ui/screens/chat/components/input.cljs#L109

can you have a check?

**Comment by Ferossgp** [comment]
@tbenr oh, sorry for that, you will need to add `:flex-direction :row` here https://github.com/status-im/status-react/blob/e8fca7ce6877272ae51832939d032f04789bde22/src/status_im/ui/screens/chat/components/style.cljs#L106

**Comment by errorists** [comment]
> Warn user if he switched to other views that recording will be not saved

I think it's okay to not warn, preferable would be to keep the recording and its state until quitting chat. We could try doing that for an Audio Messages 1.1

With regards to chat lists and notifications, @hesterbruikman what naming are we going with `Audio Message` or `Voice message`?

**Comment by tbenr** [comment]
@churik thanks :)

1) done
2) done (i put "Audio" since for images we display "Image"). I don't know how to change push notification message (@cammellos is content type missing here? https://github.com/status-im/status-react/blob/ddcb1fa54acd65ba755ac1e6ff8ddeb17c2993cb/modules/react-native-status/android/src/main/java/im/status/ethereum/module/NewMessageSignalHandler.java#L392)
3) I fixed several bugs related to back button and when sending app in background. I think this should be fixed now. Can you try again?
4) I haven't change current behaviour as per @errorists  comment

**Comment by churik** [comment]
@tbenr Thanks for the fixes!
as for issue that @simonricoo mentioned:
## ISSUE 6: recorder starts  anyway when you deny access to microphone (IOS only)
**Steps:**
1) start 1-1 chat
2) tap on audio message in input
3) deny access
At attempt to send this record got an error.
NOTE: access dialog for some reason is not shown on video, but there was popup and I tapped "Deny".
ezgif com-video-to-gif (10)

**Comment by tbenr** [comment]
@churik thank for sharing! just a confirmation: the permission dialog appear after you press the mic icon and before you press rec button, right?

**Comment by tbenr** [comment]
@Ferossgp @flexsurfer i think that we need to integrate https://github.com/react-native-community/react-native-permissions to have a better handling of permissions on ios. WDYT? If yes, I do it here or in a dedicated PR?

**Comment by Ferossgp** [comment]
@tbenr you could do it in this PR to not slow down this feature while the other PR will be reviewed separately. This seems like a small addition, and should not add much complexity here

**Comment by tbenr** [comment]
@Ferossgp @churik it should be ok now

**Comment by churik** [comment]
@tbenr 
## ISSUE 7: error is shown when putting app from background while audio is playing (Android)

**Steps:**
1) start 1-1 chat
2) send audio message
3) play audio
4) while playing, put in background and back

To continue work with app the only way is app restarting

ezgif com-video-to-gif (11)
Full logcat: [logcat-andr.log](https://github.com/status-im/status-react/files/4964552/logcat-andr.log)

**Comment by churik** [comment]
@errorists 
question: should audio recording will be hidden in input field when sending images?
Audio is sent separately, so there's no way to send audio as a comment to photo, hence it looks a bit weird.
photo_2020-07-23 13 48 51

**Comment by errorists** [comment]
@churik yes, you're absolutely right. Audio is standalone, it can't be combined with images or direct replies
Screenshot 2020-07-23 at 13 52 39

[here are the available states in Figma when in doubt](https://www.figma.com/file/aS1ct66VQ6V0cio7vSqS8UoG/Chat?node-id=10033%3A247)

**Comment by tbenr** [comment]
ok, later this evening i'll fix both

**Comment by tbenr** [comment]
@churik 
ISSUE 7: i should have fixed that. but i can't test on a physical iOS. The changes I applied have problems on iOS simulator but reading around it seems to be a simulator issue. A physical should work.

mic icon on sending image: fixed.

**Comment by flexsurfer** [comment]
awesome work, i have two comments:
1) is it possible to start recording "on press", because in other messengers you have to hold button to record, so i was doing the same but it didn't work, to start recording you have to "click"
2) is it possible to listen before sending? 
thanks

**Comment by churik** [comment]
@tbenr 
I rechecked - all issues are fixed;
Checked on Android 10, Android 8 and IOS (10, 13):
- audio messages in 1-1 and group chats
- PNs
- pairing devices
- fetching audio messages from offline
- recorder in background/foreground/navigating to different views
- playing audio while app is on background / foreground
- audio messages have not arrived from blocked user
- recorded max time audio

@flexsurfer 
comments from https://github.com/status-im/status-react/pull/10810#issuecomment-664236121 are not blockers, if I understand right and can be addressed separately?

For me it looks like ready to release

**Comment by flexsurfer** [comment]
yes not blockers just questions

**Comment by tbenr** [comment]
@churik great!
@flexsurfer 
 I left the audio player under the hood, to check audio recording and get actual duration before sending. So it is just matter of designing a convenient UI\interaction for the playback. Within this possible redesign we could think if move to press&hold or leave the "click" style. We need to consider how you are going to "pause" recording if we want to change.

**Comment by cammellos** [comment]
@tbenr again, amazing work!
I have merged the status-go code,
could you pleas update your branch to use the tag: 
`v0.56.1`

```
maks shell 
scripts/update-status-go.sh 'v0.56.1'
```

After that is ready to be merged on my side.
Thank you!

File: src/status_im/audio/core.cljs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                                 (on-pause)
+                                 (on-play)))))))
+
+(defn seek-player [player value on-seek on-error]
+  (when (and player (.-canPlay ^js player))
+    (.seek ^js player value #(if %
+                               (on-error {:error (.-err %) :message (.-message %)})
+                               (on-seek)))))
+
+(defn canPlay? [player]
```
Comment (flexsurfer) [inline src/status_im/audio/core.cljs]: `can-play?`

File: src/status_im/chat/models/input.cljs [hunk 6]
```diff
@@ -108,6 +108,15 @@
...
                                             :text         "Update to latest version to see a nice image here!"})))))
 
+(fx/defn send-audio-message
+  [cofx audio-path duration current-chat-id]
+  (when-not (string/blank? audio-path)
+    (chat.message/send-message cofx {:chat-id           current-chat-id
+                                     :content-type      constants/content-type-audio
+                                     :audio-path        audio-path
+                                     :audio-duration-ms duration
+                                     :text              "Update to latest version to listen to an audio message here!"})))
```
Comment (flexsurfer) [inline src/status_im/chat/models/input.cljs]: probably we need to translate this, not sure

File: src/status_im/events.cljs [hunk 7]
```diff
@@ -1237,7 +1242,8 @@
 (fx/defn on-going-in-background [{:keys [db now]}]
   {:db (-> db
            (dissoc :app-active-since)
-           (assoc :app-in-background-since now))})
+           (assoc :app-in-background-since now))
+   :dispatch-n '([:audio-recorder/on-background] [:audio-message/on-background])})
```
Comment (flexsurfer) [inline src/status_im/events.cljs]: just wondering why list?

File: src/status_im/audio/core.cljs (hunk 5 above)
Comment (tbenr) [reply src/status_im/audio/core.cljs]: oh yes..

File: src/status_im/events.cljs (hunk 7 above)
Comment (tbenr) [reply src/status_im/events.cljs]: @flexsurfer you're concerned because i fire two events or because it is a list?

Comment (flexsurfer) [reply src/status_im/events.cljs]: last one :) usually vector is used

**Comment by flexsurfer** [comment]
not sure if ios fail related to this PR, i would merge it, but let's wait for Jakub

File: src/status_im/events.cljs (hunk 7 above)
Comment (tbenr) [reply src/status_im/events.cljs]: don't even remember from where i took that :) i'll change it later

**Comment by tbenr** [comment]
@flexsurfer noticed that lately ios build sometime fails. shortly I'll push your suggestions and let's see if builds.

**Comment by churik** [comment]
@tbenr @flexsurfer should I retest briefly?

**Comment by tbenr** [comment]
@flexsurfer fixed all.
@churik don't think so, changes are minimal. Another check from @flexsurfer should be enough

**Comment by flexsurfer** [comment]
we could run e2e just to make sure, no manual testing needed

**Comment by tbenr** [comment]
I need you @cammellos to resolve the status-go conflict :-)

**Comment by cammellos** [comment]
@tbenr you can keep the version in this branch (v0.56.1) as it includes the previous changes

**Comment by cammellos** [comment]
> 2. test_pair_devices_sync_one_to_one_contacts_public_chat


Is a bug in status-go, not related to this PR and will be fixed in a separate PR  https://github.com/status-im/status-go/pull/2006

**Comment by churik** [comment]
other failures are not related to this PR. Ready to merge @cammellos

**Comment by jinhojang6** [comment]
Fantastic work! @tbenr
